# 2021/05/30 小賴老師上課筆記

## Javascript 的 Callback, Promise, Async / Await

### callback function

**callback function 不一定是非同步，但是要非同步必須用到 callback function**

#### 需要按照先後順序處理的工作

- 註冊帳號: 先驗證 -> 建立新帳號
- 以下的 morning routine 作為範例  
  在非同步的程式語言中，會用到 callback function 內包含 callback function 來解決必須按照順許處理的工作  
  在下面範例中，在 stack 中執行刷牙工作時，其中的 cb function 會被排入 queue 中，等待刷牙工作執行完成會被執行，如此一直重複直到最內層的吃午餐工作被執行完畢

```javascript
let doWork = function (job, timer, cb) {
  setTimeout(() => {
    let dt = new Date();
    cb(null, `完成工作: ${job} at ${dt.toISOString()}`);
  }, timer);
};

let dt = new Date();
console.log(`開始工作 at ${dt.toISOString()}`);
doWork("刷牙", 2000, function (err, result) {
  if (err) {
    console.error(err);
    return;
  }
  console.log(result);
  doWork("吃早餐", 3000, function (err, result) {
    if (err) {
      console.error(err);
      return;
    }
    console.log(result);

    doWork("寫功課", 5000, function (err, result) {
      if (err) {
        console.error(err);
        return;
      }
      console.log(result);

      doWork("吃午餐", 3000, function (err, result) {
        if (err) {
          console.error(err);
          return;
        }
        console.log(result);
      });
    });
  });
});
```

**以上的問題是會產生 callback hell，讓程式變得不易閱讀和維護**，要解決以上的問題可以使用**Promise**

### Promise

Promise 是一個非同步運算的最終完成或失敗的物件，他讓非同步方法回傳值的方式很像同步方法

Promise 代理一個建立時，不用預先得知的值，並由人定義成功與失敗的狀態，以及成功時(resolved)的成功值(success value)和拒絕時(reject)的錯誤訊息(failure reason)

```
new Promise( /* executor */ function(resolve, reject) {
  resolve("回傳結果");
  reject("回傳失敗")
 } );
```

executor 會在建構式回傳物件實體前立即執行，也就是說當傳入這個函式時，Promise 物件會立即決定裡面的狀態，決定要執行 resolve 或是 reject 函式。

Promise 一共有三種狀態

- pending (初始狀態)
- fulfilled (實現)
- rejected (拒絕)

當 promise fulfilled 時，就會呼叫 resolve 進入.then，失敗的時候會呼叫 reject 進入.catch
<br/>

#### Promise 的使用方法

1. 先 new 一個 promise
2. 給予 promise 物件內的 resolve 函式 success value，給予 reject 函式 failure reason
3. 在.then 中以 function 處理成功值
4. 在.catch 中以 function 處理錯誤訊息
   <br/>

#### Promise 的串接

![](https://cdn.rawgit.com/Vectaio/a76330b025baf9bcdf07cb46e5a9ef9e/raw/26c4213a93dee1c39611dcd0ec12625811b20a26/js-promise.svg)

如果在.then 中 return secondPromise，則可以在下一個.then 中處理 secondPromise 的成功值，由此可將多個 promise 串接起來。

```javascript
Promise((resolve, reject) => {
  resolve(successValue);
  reject(failureReason);
})
  .then((result) => {
    return secondPromise(result);
  })
  .then((secondResult) => {
    console.log(secondResult);
  })
  .catch((failureReason) => {
    console.log("Promise和secondPromise的錯誤訊息: ", failureReason);
  });
```

#### Promise 的其他方法

**Promise.all** 等到全部實現（或一個拒絕）。
所有工作並行的做(非同步)

```javascript
let doWorkPromise = function (job, timer, success) {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      let dt = new Date();
      if (success) {
        // 成功
        return resolve(`完成工作: ${job} at ${dt.toISOString()}`);
      }
      reject(`!!工作失敗: ${job} at ${dt.toISOString()}`);
    }, timer);
  });
};

let dt = new Date();
console.log(`開始工作 at ${dt.toISOString()}`);
let count = 0;

let p1 = doWorkPromise("寫功課", 3000, true);
let p2 = doWorkPromise("聽音樂", 2500, false);
let p3 = doWorkPromise("同學聊天", 3500, true);

// 1. 並行地做 （非同步）
// 2. 三件事「都」做完的時候通知我 -> console.log
Promise.all([p1, p2, p3])
  .then((values) => {
    console.log("race");
    console.log(values);
  })
  .catch((err) => {
    console.log("錯誤！", err);
  });

//print "完成工作: 寫功課 at 2021-06-01T07:02:51.415Z", "完成工作: 聽音樂 at 2021-06-01T07:02:50.914Z", "完成工作: 同學聊天 at 2021-06-01T07:02:51.915Z"
```

**Promise.race**
回傳最快實現的 promise 物件  
所有工作並行的做(非同步)

```javascript
Promise.race([p1, p2, p3])
  .then((values) => {
    console.log(values);
  })
  .catch((err) => {
    console.log("錯誤！", err);
  });

//print 完成工作: 聽音樂 at 2021-06-01T07:03:35.443Z
```

### Async / Awiat

- async / await based on Promise --> 必須先 new 一個 promise
- 相依的工作會用到此方法
- await 會**等**你呼叫 resolve，得到 result 後再往下一行
- await 必須包在 async function 中才能執行
- 因為沒有 promise 的.catch，所以會用 try{ ... }catch(err){ ... }來處理錯誤訊息

```javascript
let doWorkPromise = function (job, timer, success) {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      let dt = new Date();
      if (success) {
        // 成功
        return resolve(`完成工作: ${job} at ${dt.toISOString()}`);
      }
      reject(`!!工作失敗: ${job} at ${dt.toISOString()}`);
    }, timer);
  });
};
// 刷完牙 > 吃早餐 > 寫功課
// 1. await 只能用在 async function 裡
async function manager() {
  try {
    let result = await doWorkPromise("刷完牙", 2000, true);
    console.log("await", result);

    result = await doWorkPromise("吃早餐", 2500, true);
    console.log("await", result);

    result = await doWorkPromise("寫功課", 3000, true);
    console.log("await", result);
  } catch (err) {
    console.log("錯誤", err);
  } finally {
    console.log("最後做完了");
  }
}
```

### 參考文章

[Promise 物件建立與基本使用](http://promise.eddychang.me/docs/contents/ch4_basic_usage/)  
[MDN Promise](https://developer.mozilla.org/zh-TW/docs/Web/JavaScript/Reference/Global_Objects/Promise)
